rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,453 @@
+            Teleport Proxy->>Local Machine: 
+        end
+        Local Machine->>Teleport Proxy: get headless login request info<br/>(rpc GetHeadlessLoginRequest)
+        Teleport Proxy->>Local Machine: 
+        Note over Local Machine: share request details with user and<br/>prompt for confirmation (y/N)
+        Local Machine->>Teleport Proxy: initiate headless request approval<br/>(rpc UpdateHeadlessLoginRequestState)
+        Teleport Proxy->>Local Machine: MFA challenge
+        Note over Local Machine: user taps YubiKey
+        Local Machine->>Teleport Proxy: signed MFA challenge response
+        Teleport Proxy->>Teleport Auth: update request state
+    end
+
+    Teleport Auth->>-Teleport Proxy: unblock on request state change
+    Teleport Proxy->>Teleport Auth: Request signed user certificates
+    Teleport Auth->>Teleport Proxy: Issue signed user certificates
+    Teleport Proxy->>Remote Machine: user certificates<br/>(MFA-verified, 1 minute TTL)
+    Note over Remote Machine: Connect to user@node01
+```
+
+This flow can be broken down into three parts: headless login initiation, local authentication, and certificate retrieval.
+
+#### Headless login initiation
+
+First, the client initiates headless login, providing a request UUID, the command being requested (e.g. `tsh ssh user@host`), and normal login parameters including the client public key.
+
+```go
+type HeadlessLoginRequest struct {
+    SSHLogin
+    // User is a teleport username
+    User string `json:"user"`
+    // RequestID is a uuid for the request
+    RequestID string
+    // Command is the client command being requested
+    Command string
+}
+
+// SSHLogin contains common SSH login parameters.
+type SSHLogin struct {
+    // ProxyAddr is the target proxy address
+    ProxyAddr string
+    // PubKey is SSH public key to sign
+    PubKey []byte
+    ...
+}
+```
+
+These request details are saved on the Auth server in a `HeadlessLoginRequest` object, with the backend prefix `/headless_login_requests/`. They will be saved with a 1 minute TTL, by which point the user should have completed the authentication flow. The request will begin in the pending state, to be later approved/denied by the user locally.
+
+```go
+type HeadlessLoginRequest struct {
+    ID string
+    // User is a teleport username
+    User string
+    // PubKey is SSH public key to sign
+    PubKey []byte
+    // Command is the client command being requested
+    Command string
+    // State is the request state (pending, approved, or denied)
+    State string
+}
+```
+
+#### Local authentication
+
+In parallel to making the headless login request above, the client will generate a command and URL to initiate local authentication to approve the request: `tsh headless approve <request_id>` and `https://proxy.example.com/headless/<request_id>/approve`. The command and URL are shared with the user so they can complete local authentication for the request in a local terminal or web browser.
```

Comment:
Nit: Maybe it's just me but "tsh headless approve" rubs me the wrong way a little bit. Feels like the CLI command exposes the implementation details. Maybe "tsh session approve"? Cause technically we're approving a single resource session. I guess it's more of a @klizhentas and @xinding33 question.